cmd/kube-controller-manager/app/controllermanager.go

```diff
@@ -389,6 +397,12 @@ func CreateControllerContext(s *config.CompletedConfig, rootClientBuilder, clien
 		return ControllerContext{}, fmt.Errorf("failed to wait for apiserver being healthy: %v", err)
 	}
 
+	// Use a discovery client capable of being refreshed.
+	discoveryClient := rootClientBuilder.ClientOrDie("controller-discovery")
```

Comment:
You have the stop channel here.  Move the `.Reset()` call from the `GarbageCollection.Sync` method to a gofunc you start here with that channel to stop it.

Reply:
ok. also **reset** with the same period as  `GarbageCollection.Sync`